sky/server/middleware_utils.py

```diff
@@ -0,0 +1,148 @@
+            if decision == WebSocketDecision.ACCEPT:
+                await self.app(scope, receive, send)
+            elif decision == WebSocketDecision.UNAUTHORIZED:
+                await send({
+                    'type': 'websocket.close',
+                    'code': 4401,
+                    'reason': 'Unauthorized',
+                })
+            elif decision == WebSocketDecision.FORBIDDEN:
+                await send({
+                    'type': 'websocket.close',
+                    'code': 4403,
+                    'reason': 'Forbidden',
+                })
+            else:
+                await send({
+                    'type': 'websocket.close',
+                    'code': 1011,
+                    'reason': 'Internal Server Error',
+                })
+
+        async def _run_websocket_dispatch(self, scope) -> WebSocketDecision:
+            http_scope = self._build_http_scope(scope)
+            http_receive = self._http_receive_adapter()
+            request = fastapi.Request(http_scope, receive=http_receive)
+            call_next_called = False
+            stub_response = fastapi.Response(status_code=http.HTTPStatus.OK)
+
+            async def call_next(req):
+                del req
+                nonlocal call_next_called
+                call_next_called = True
+                return stub_response
+
+            try:
+                response = await self.dispatch(request, call_next)
+            except Exception as e:  # pylint: disable=broad-except
+                logger.error('Exception occurred in middleware dispatch for '
+                             f'WebSocket scope: {e}')
+                return WebSocketDecision.ERROR
+
+            if response is None:
+                response = stub_response
+
+            status_code = response.status_code
+
+            if call_next_called and 200 <= status_code < 400:
+                return WebSocketDecision.ACCEPT
+            if status_code == http.HTTPStatus.UNAUTHORIZED:
+                return WebSocketDecision.UNAUTHORIZED
+            if status_code == http.HTTPStatus.FORBIDDEN:
+                return WebSocketDecision.FORBIDDEN
+            return WebSocketDecision.ERROR
+
+        @staticmethod
+        def _build_http_scope(scope):
+            state = scope.setdefault('state', {})
+            scheme = scope.get('scheme', 'ws')
+            if scheme == 'ws':
+                http_scheme = 'http'
+            elif scheme == 'wss':
+                http_scheme = 'https'
+            else:
+                http_scheme = scheme
+            http_scope = dict(scope)
+            http_scope['type'] = 'http'
+            http_scope['scheme'] = http_scheme
+            http_scope['method'] = 'GET'
+            http_scope['http_version'] = scope.get('http_version', '1.1')
+            http_scope['state'] = state
+            return http_scope
+
+        @staticmethod
+        def _http_receive_adapter():
```

Comment:
```suggestion
        def _http_receive_adapter() -> 'starlette.types.Receive':
```
does this work? Feel free to ignore